odoo/tests/common.py

```diff
@@ -1633,82 +1646,136 @@ def request(self, *args, **kwargs):
+    def raise_for_status(self):
+        reason = self.status
+        if 400 <= self.status_code < 500:
+            http_error_msg = (
+                f"{self.status_code} Client Error: {reason} for url: {self.url}"
+            )
+        elif 500 <= self.status_code < 600:
+            http_error_msg = (
+                f"{self.status_code} Server Error: {reason} for url: {self.url}"
+            )
+        else:
+            http_error_msg = ""
+
+        if http_error_msg:
+            raise HTTPError(http_error_msg, response=self)
+    @property
+    def ok(self):
+        with contextlib.suppress(HTTPError):
+            self.raise_for_status()
+            return True
+        return False
+
+    @property
+    def text(self):
+        return self.get_data(as_text=True)
+
+    # werkzeug 2 adds a `json` to `Response`, but it's a property...
+    # pylint: disable=invalid-overridden-method
+    def json(self):
+        return json.loads(self.text)
+
+    @property
+    def content(self):
+        return self.data
+
+    @property
+    def cookies(self):
+        cookies = BaseCookie()
+        for h, v in self.headers.items():
+            if h == 'Set-Cookie':
+                cookies.load(v)
+        return {name: morsel.value for name, morsel in cookies.items()}
+
+class TestClientOpener(werkzeug.test.Client):
+    def __init__(self, cr):
+        super().__init__(odoo.http.root, response_wrapper=TestResponse)
+        self.cr = cr
+
+    def open(self, *args, **kwargs):
+        self.cr.flush()
+        self.cr.clear()
+        assert not kwargs.pop('files', None), \
+            "TODO: translate files between requests and werkzeug APIs"
+        kwargs.pop('timeout', None)
+        # translate kwargs from requests
+        kwargs['follow_redirects'] = kwargs.pop('allow_redirects', False)
+
+        base_url = f"http://{HOST}:{odoo.tools.config['http_port']}"
+        if not args[0].startswith(('/', base_url + '/')):
+            raise AssertionError(f"WSGI client implies {HOST}")
+        args = (
+            args[0].replace(base_url, ''),
+            *args[1:]
+        )
 
+        kwargs['environ_base'] = {
+            # defaults the current IP as the remote (gethostbyname is probably
+            # unnecessary but...)
+            'REMOTE_ADDR': gethostbyname(HOST),
+        }
+        return super().open(*args, **kwargs)
+
+    @property
+    def cookies(self):
+        # should be a RequestsCookieJar, which is a CookieJar subclass with a
+        # complex MutableMapping interface (and some more), but just implement
+        # setitem for now, client has a cookie_jar attribute which is its own
+        # TestCookieJar
+        return CookiesProxy(self)
+class CookiesProxy:
+    def __init__(self, client):
+        self.client = client
+
+    def __setitem__(self, key, value):
+        # nb: that's not the cookiejar method
+        self.client.set_cookie("", key, value)
+
+class QueryCase(TransactionCase):
```

Comment:
We should make that class private. I'm not fan of the name, to me "query" is too related to sql/databases. What's about `_WebCase`?

Reply:
I think of web as the layer "above" http, not below it (to say nothing of below wsgi).

Reply:
Also this stuff is probably not done and half the object should be hidden. I think it would make sense to move all the test cases into submodules and only re-export *some* from here.